src/details/ArborX_DetailsDendrogram.hpp

```diff
@@ -0,0 +1,158 @@
+#include <ArborX_DetailsKokkosExtViewHelpers.hpp>
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUnionFind.hpp>
+#include <ArborX_MinimumSpanningTree.hpp> // WeightedEdge
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX::Details
+{
+
+// Compute permutation that orders edges in increasing order
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<unsigned int *, MemorySpace>
+computeEdgesPermutation(ExecutionSpace const &exec_space,
+                        Kokkos::View<WeightedEdge *, MemorySpace> &edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::dendrogram::sort_edges");
+
+  int const num_edges = edges.size();
+
+  Kokkos::View<float *, MemorySpace> weights(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::weights"),
+      num_edges);
+  Kokkos::parallel_for(
+      "ArborX::Dendrogram::copy_weights",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
+      KOKKOS_LAMBDA(int const e) { weights(e) = edges(e).weight; });
+
+  auto permute = Details::sortObjects(exec_space, weights);
+
+  Kokkos::Profiling::popRegion();
+
+  return permute;
+}
+
+template <typename Edges, typename UnionFind, typename Permute,
+          typename Parents>
+void dendrogramUnionFindHost(Edges edges_host, Permute permute,
+                             UnionFind union_find, Parents &parents_host)
+{
+  Kokkos::DefaultHostExecutionSpace exec_space;
+
+  int const num_edges = edges_host.size();
+  int const num_vertices = num_edges + 1;
+  auto const vertices_offset = num_edges;
+
+  constexpr int UNDEFINED = -1;
+
+  Kokkos::View<int *, Kokkos::HostSpace> set_edges_host(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::labels"),
+      num_vertices + num_edges);
+  Kokkos::deep_copy(exec_space, set_edges_host, UNDEFINED);
+
+  for (int k = 0; k < num_edges; ++k)
+  {
+    auto e = permute(k);
+
+    int i = edges_host(e).source;
+    int j = edges_host(e).target;
+
+    for (int s : {i, j})
+    {
+      auto child = set_edges_host(union_find.representative(s));
+      if (child == UNDEFINED)
+      {
+        // This will happen when a vertex has not been assigned a parent yet
+        parents_host(vertices_offset + s) = e;
+      }
+      else
+      {
+        parents_host(child) = e;
+      }
+    }
+
+    union_find.merge(i, j);
+
+    set_edges_host(union_find.representative(i)) = e;
+  }
+  parents_host(permute(num_edges - 1)) = UNDEFINED; // root
+}
+
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<int *, MemorySpace>
+dendrogramUnionFind(ExecutionSpace const &exec_space,
+                    Kokkos::View<WeightedEdge *, MemorySpace> edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::dendrogram_union_find");
+
+  Kokkos::Profiling::pushRegion(
+      "ArborX::Dendrogram::dendrogram_union_find::edge_sort");
```

Comment:
`computeEdgesPermutation` is wrapped in his own profiling region.  It this not a bit excessive?